pkg/controller/replicaschedulingpreference/controller.go

```diff
@@ -15,6 +14,709 @@ See the License for the specific language governing permissions and
+func StartReplicaSchedulingPreferenceController(fedConfig, kubeConfig, crConfig *restclient.Config, stopChan <-chan struct{}, minimizeLatency bool) error {
+	userAgent := "replicaschedulingpreference-controller"
+	restclient.AddUserAgent(fedConfig, userAgent)
+	fedClient := fedclientset.NewForConfigOrDie(fedConfig)
+	restclient.AddUserAgent(kubeConfig, userAgent)
+	kubeClient := kubeclientset.NewForConfigOrDie(kubeConfig)
+	restclient.AddUserAgent(crConfig, userAgent)
+	crClient := crclientset.NewForConfigOrDie(crConfig)
+	controller, err := newReplicaSchedulingPreferenceController(fedConfig, fedClient, kubeClient, crClient)
+	if err != nil {
+		return err
+	}
+	if minimizeLatency {
+		controller.minimizeLatency()
+	}
+	glog.Infof(fmt.Sprintf("Starting replicaschedulingpreferences controller"))
+	controller.Run(stopChan)
+	return nil
+}
+
+// newReplicaSchedulingPreferenceController returns a new ReplicaSchedulingPreference Controller for the given client
+func newReplicaSchedulingPreferenceController(fedConfig *restclient.Config, fedClient fedclientset.Interface, kubeClient kubeclientset.Interface, crClient crclientset.Interface) (*ReplicaSchedulingPreferenceController, error) {
+	broadcaster := record.NewBroadcaster()
+	broadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: kubeClient.CoreV1().Events("")})
+	recorder := broadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: fmt.Sprintf("replicaschedulingpreference-controller")})
+
+	s := &ReplicaSchedulingPreferenceController{
+		reviewDelay:             time.Second * 10,
+		clusterAvailableDelay:   time.Second * 20,
+		clusterUnavailableDelay: time.Second * 60,
+		smallDelay:              time.Second * 3,
+		updateTimeout:           time.Second * 30,
+		workQueue:               workqueue.New(),
+		backoff:                 flowcontrol.NewBackOff(5*time.Second, time.Minute),
+		eventRecorder:           recorder,
+		fedClient:               fedClient,
+	}
+
+	// Build delivereres for triggering reconciliations.
+	s.deliverer = util.NewDelayingDeliverer()
+	s.clusterDeliverer = util.NewDelayingDeliverer()
+
+	// Start informers on the resources for the federated type
+	deliverObj := func(obj pkgruntime.Object) {
+		s.deliverObj(obj, 0, false)
+	}
+
+	s.templateStore, s.templateController = cache.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (pkgruntime.Object, error) {
+				return fedClient.FederatedschedulingV1alpha1().ReplicaSchedulingPreferences(metav1.NamespaceAll).List(options)
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				return fedClient.FederatedschedulingV1alpha1().ReplicaSchedulingPreferences(metav1.NamespaceAll).Watch(options)
+			},
+		},
+		&fedschedulingv1a1.ReplicaSchedulingPreference{},
+		util.NoResyncPeriod,
+		util.NewTriggerOnAllChanges(deliverObj),
+	)
+
+	s.templateStore, s.templateController = cache.NewInformer(
```

Comment:
Is this actually ready for review?  This line and 161 would appear to conflict and cause me to question whether this controller is functional.  Are you intending to add tests to verify the proposed functionality?

Reply:
I did an initial verification of api availability only earlier; have verified the complete flow of overrides, placement and actual deployment being created in fcp and cluster respectively after a replicaschedulingpreference and template is created in fcp, today.
Tests us what I am working on now and intend to submit that as another PR.

Reply:
I think minimal automated testing is a prereq for reviewing, let alone merging.  It doesn't have to be exhaustive by any means, but I would like to be reviewing code that had some automatic validation in place.  I'm not a very good runtime.

Reply:
:-); I will update some tests as part of this PR.

Reply:
@marun updated, took me some time to absorb and rebase on your recent changes. PTAL.
@shashidharatd @font if possible, you can also pitch in for review please.